### PR TITLE
docs: Plugin flat config migration guide

### LINKED ISSUE #17242: Change Request: document flat configs in plugins
### ESLint version

8.42.0

### What problem do you want to solve?

Current [documentation page on writing plugins](https://eslint.org/docs/latest/extend/plugins) uses legacy configs

### What do you think is the correct solution?

It should provide a useful example of how a flat config can be included in a plugin without circular imports.

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Added a migration guide for plugins to implement flat config. This focuses on the `configs` key but also touches on everything else that plugins can export.

Fixes #17242

#### Is there anything you'd like reviewers to focus on?

Did I miss any details? Do the examples make sense?

ping @bradzacher @JoshuaKGoldberg

### CODE REVIEW

**Review by bradzacher — COMMENTED** [review]
The thing that's missing from my POV is how a plugin/config package should support both `.eslintrc` and flat configs.

Whilst, yes, v9 will switch the default - I have no doubt that there will be a long-tail of users that take a long time to (a) upgrade to v9 and (b) migrate to flat configs.

If we switch our configs over to only support flat configs then the majority of our users will be "left in the lurch" and thus will be stuck.

For example I know that at Canva or at Meta - migrating the eslint config would likely be a multi-week project given how many overrides there are - it will take a lot of testing to validate the migration is 1:1. Such a project requires prioritisation and scheduling which can delay it for quite some time.

So to summarise - I think these docs need to answer the question "how can I support flat configs and legacy configs simultaneously?"

File: docs/src/extend/plugin-migration-flat-config.md [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+1. **Export a CommonJS entrypoint.** The old configuration system cannot load plugins that are published only in ESM format. If your source code is in ESM, then you'll need to use a bundler that can generate a CommonJS version and use the [`exports`](https://nodejs.org/api/packages.html#package-entry-points) key in your `package.json` file to ensure the CommonJS version can be found by Node.js.
+1. **Keep the `environments` key.** If your plugin exports custom environments, you should keep those as they are and also export the equivalent flat configs as described above. The `environments` key is ignored when ESLint is running in flat config mode.
+1. **Export both eslintrc and flat configs.** The `configs` key is only validated when a config is used, so you can provide both formats of configs in the `configs` key. We recommend that you append older format configs with `-legacy` to make it clear that these configs will not be supported in the future. For example, if your primary config is called `recommended` and is in flat config format, then you can also have a config named `recommended-legacy` that is the eslintrc config format.
```
Comment (bradzacher) [inline docs/src/extend/plugin-migration-flat-config.md:287]: I guess delineating the config name with `-legacy` is probably the absolute best way to ensure that you can keep the backwards compat and cleanly define both configs without muddling the new vs old keys and such.

**Review by bradzacher — COMMENTED** [review]
that's looking good to me - thanks for the clarification!

**Review by mdjermanovic — APPROVED** [review]
LGTM, just one typo.

**Comment by jaydenseric** [comment]
Many legacy ESLint plugins published to npm don't contain types because the authors didn't expect people to be importing the plugins and using them in code. TypeScript users will now greatly appreciate the published plugin packages having types, so they can be imported and used in project code with type safety. It would be good to add to the migration guide that flat config plugins should be typed (if you want to support TypeScript projects), and how to type them.
